test/test_optional.cpp

```diff
@@ -0,0 +1,697 @@
+            optional_double d1(initial);
+            optional_double d2(expected);
+            optional_double empty(std::nullopt);
+
+            swap(d1, d2);
+            CHECK_EQ(d1.value(), expected);
+            CHECK_EQ(d2.value(), initial);
+
+            swap(d1, empty);
+            CHECK_EQ(empty.value(), expected);
+            CHECK_FALSE(d1.has_value());
+        }
+
+        TEST_CASE("reset")
+        {
+            constexpr double initial = 1.2;
+            optional_double d(initial);
+            d.reset();
+            CHECK_FALSE(d.has_value());
+        }
+
+        TEST_CASE("equality comparison")
+        {
+            constexpr double initial = 1.2;
+            constexpr double other = 2.5;
+
+            optional_double d1(initial);
+            optional_double d2(other);
+            optional_double empty;
+
+            CHECK(d1 == d1);
+            CHECK(d1 == d1.value());
+            CHECK(d1 != d2);
+            CHECK(d1 != d2.value());
+            CHECK(d1 != empty);
+            CHECK(empty == empty);
+        }
+
+        TEST_CASE("inequality comparison")
+        {
+            constexpr double initial = 1.2;
+            constexpr double other = 2.5;
+
+            optional_double d1(initial);
+            optional_double d2(other);
+            optional_double empty;
+
+            // opearator <=
+            CHECK(d1 <= d1);
+            CHECK(d1 <= d1.value());
+            CHECK(d1 <= d2);
+            CHECK(d1 <= d2.value());
+            CHECK_FALSE(d2 <= d1);
+            CHECK_FALSE(d2 <= d1.value());
+            CHECK(empty <= d1);
+            CHECK_FALSE(d1 <= empty);
+
+            // operator >=
+            CHECK(d1 >= d1);
+            CHECK(d1 >= d1.value());
+            CHECK(d2 >= d1);
+            CHECK(d2 >= d1.value());
+            CHECK_FALSE(d1 >= d2);
+            CHECK_FALSE(d1 >= d2.value());
+            CHECK(d1 >= empty);
+            CHECK_FALSE(empty >= d1);
+
+            // operator<
+            CHECK_FALSE(d1 < d1);
+            CHECK_FALSE(d1 < d1.value());
+            CHECK(d1 < d2);
+            CHECK(d1 < d2.value());
+            CHECK(empty < d1);
+            CHECK_FALSE(d1 < empty);
+
+            // oprator>
+            CHECK_FALSE(d1 > d1);
+            CHECK_FALSE(d1 > d1.value());
+            CHECK(d2 > d1);
+            CHECK(d2 > d1.value());
+            CHECK(d1 > empty);
+            CHECK_FALSE(empty > d1);
+        }
+
+        TEST_CASE("make_optional")
+        {
+            double value = 2.5;
+            auto opt = make_optional(std::move(value), true);
+            static_assert(std::same_as<std::decay_t<decltype(opt)>, optional_double>);
+            REQUIRE(opt.has_value());
+            CHECK_EQ(opt.value(), value);
+        }
+    }
+
+    using optional_proxy = optional<double&, bool>;
+
+    TEST_SUITE("optional proxy")
+    {
+        TEST_CASE("constructors")
+        {
+            double val = 1.2;
```

Comment:
Set all the constants evaluated against as `const` (in all tests that have them)